src/porepy/numerics/ad/operator_functions.py

```diff
@@ -109,59 +109,127 @@ def __repr__(self) -> str:
 
-    def __add__(self, other):
-        raise RuntimeError("AD Operator functions are meant to be called, not added.")
+    def __pow__(self, other: Operator) -> Operator:
+        raise TypeError(
+            "Operator functions must be called before applying any operation."
+        )
+
+    def __rpow__(self, other: Operator) -> Operator:
+        raise TypeError(
+            "Operator functions must be called before applying any operation."
+        )
 
-    def __sub__(self, other):
-        raise RuntimeError(
-            "AD Operator functions are meant to be called, not subtracted."
+    def __matmul__(self, other: Operator) -> Operator:
+        raise TypeError(
+            "Operator functions must be called before applying any operation."
         )
 
-    def __rsub__(self, other):
-        raise RuntimeError(
-            "AD Operator functions are meant to be called, not subtracted."
+    def __rmatmul__(self, other):
+        raise TypeError(
+            "Operator functions must be called before applying any operation."
         )
 
-    def __div__(self, other):
-        raise RuntimeError("AD Operator functions are meant to be called, not divided.")
+    def parse(self, mdg: pp.MixedDimensionalGrid):
+        """Operator functions return themselves to give the recursion in
+        :class:`~porepy.numerics.ad.operators.Operator` access to the underlying
+        :meth:`func`."""
+        return self
 
-    def __truediv__(self, other):
-        raise RuntimeError("AD Operator functions are meant to be called, not divided.")
+    def func(self, *args: FloatType) -> float | np.ndarray | AdArray:
+        """The underlying numerical function which is represented by this operator
+        function.
 
-    def parse(self, md: pp.MixedDimensionalGrid):
-        """Parsing to a numerical value.
+        Called during parsing with the numerical representation of operator arguments
+        and returning the numerical value and derivative of this operator instance.
 
-        The real work will be done by combining the function with arguments, during
-        parsing of an operator tree.
+        The numerical function calls in any case :meth:`get_values` with ``*args``.
+        If ``*args`` contains an Ad array, it calls also :meth:`get_jacobian`.
 
         Parameters:
-            md: Mixed-dimensional grid.
+            *args: Numerical representation of the operators with which this instance
+                was called. The arguments will be in the same order as the operators
+                passed to the call to this instance.
 
         Returns:
-            The instance itself.
+            If ``*args`` contains only numpy arrays, it returns the result of
+            :meth:`get_values`.
+            If it contains an Ad array, it combines the results of :meth:`get_values`
+            and :meth:`get_jacobian` in an Ad array and returns it.
 
         """
-        return self
+
+        values = self.get_values(*args)
```

Comment:
The split is done for the most general case, where values and derivatives are obtained independently.
For the case where the actual AD is implemented for example in `Function`, there is no split because that class assumes that the callables can handle AdArrays.
The more general cases include the interpolation table, or the secondary operator which will follow after this merger.